src/cpp/rtps/writer/StatefulWriter.cpp

```diff
@@ -354,32 +354,35 @@ void StatefulWriter::send_any_unsent_changes()
             for (std::unique_ptr<FlowController>& controller : mp_RTPSParticipant->getFlowControllers())
                 (*controller)(relevantChanges);
 
-            RTPSMessageGroup group(mp_RTPSParticipant, this, RTPSMessageGroup::WRITER, m_cdrmessages);
+            RTPSMessageGroup group(mp_RTPSParticipant, this, m_cdrmessages, *this);
             uint32_t lastBytesProcessed = 0;
 
             while (!relevantChanges.empty())
             {
                 RTPSWriterCollector<ReaderProxy*>::Item changeToSend = relevantChanges.pop();
-                std::vector<GUID_t> remote_readers;
-                std::vector<LocatorList_t> locatorLists;
                 bool expectsInlineQos = false;
+                locator_selector_.reset(false);
 
                 for (const ReaderProxy* remoteReader : changeToSend.remoteReaders)
                 {
-                    remote_readers.push_back(remoteReader->guid());
-                    locatorLists.push_back(remoteReader->remote_locators());
+                    locator_selector_.enable(remoteReader->guid());
                     expectsInlineQos |= remoteReader->expects_inline_qos();
                 }
 
+                if (locator_selector_.state_has_changed())
+                {
+                    group.flush_and_reset();
+                    network.select_locators(locator_selector_);
```

Comment:
Suggestion: this code is repeated and may aim to commit errors. Maybe can be put in one single location.

Reply:
It will be useful to describe the usage of `flush_and_reset` using doxygen.